crates/viewer/re_viewer_context/src/drag_and_drop.rs

```diff
@@ -13,166 +51,286 @@ use re_ui::{
-            Self::Invalid => "invalid selection".fmt(f),
+            Self::Invalid => {}
         }
+
+        item_counter.fmt(f)
     }
 }
 
-/// Display the currently dragged payload as a pill in the UI.
+#[derive(Default, Debug, Clone, Copy, PartialEq, Eq)]
+pub enum DragAndDropFeedback {
+    /// The payload type is irrelevant to me.
+    ///
+    /// For example, dropping a view and/or contain onto an existing view in the viewport is
+    /// irrelevant.
+    ///
+    /// This is the default displayed feedback, unless explicitly set otherwise by some UI hovered
+    /// UI element.
+    #[default]
+    Ignore,
+
+    /// The payload type is acceptable and could successfully be dropped at the current location.
+    Accept,
+
+    /// The payload type is correct, but it's content cannot be accepted by the current drop location.
+    ///
+    /// For example, a view might reject an entity because it already contains it.
+    Reject,
+}
+
+/// Helper to handle drag-and-drop operations.
 ///
-/// This should be called once per frame.
-pub fn drag_and_drop_payload_cursor_ui(ctx: &egui::Context) {
-    if let Some(payload) = egui::DragAndDrop::payload::<DragAndDropPayload>(ctx) {
-        if let Some(pointer_pos) = ctx.pointer_interact_pos() {
-            let icon = match payload.as_ref() {
-                DragAndDropPayload::Contents { .. } => &re_ui::icons::DND_MOVE,
-                // don't draw anything for invalid selection
-                DragAndDropPayload::Invalid => return,
-            };
-
-            let layer_id = egui::LayerId::new(
-                egui::Order::Tooltip,
-                egui::Id::new("drag_and_drop_payload_layer"),
-            );
-
-            let mut ui = egui::Ui::new(
-                ctx.clone(),
-                egui::Id::new("rerun_drag_and_drop_payload_ui"),
-                egui::UiBuilder::new().layer_id(layer_id),
-            );
-
-            ui.set_opacity(0.7);
-
-            let response = drag_pill_frame(matches!(
-                payload.as_ref(),
-                &DragAndDropPayload::Invalid { .. }
-            ))
-            .show(&mut ui, |ui| {
-                let text_color = ui.visuals().widgets.inactive.text_color();
-
-                ui.horizontal(|ui| {
-                    ui.spacing_mut().item_spacing.x = 2.0;
-
-                    ui.small_icon(icon, Some(text_color));
-                    ui.label(egui::RichText::new(payload.to_string()).color(text_color));
-                });
-            })
-            .response;
+/// This helper should be constructed at the beginning of the frame and disposed of at the end.
+/// Its [`Self::payload_cursor_ui`] method should be called late during the frame (after the rest of
+/// the UI has a chance to update the feedback).
+pub struct DragAndDropManager {
+    /// Items that may not be dragged, e.g., because they are not movable nor copiable.
+    undraggable_items: ItemCollection,
+
+    feedback: Arc<Mutex<DragAndDropFeedback>>,
```

Comment:
oOo I've been wanting something like this for exactly this case. Surely that "pure" enum ends up being represented by an atomic type, so that's a nice improvement.

Reply:
> Surely that "pure" enum ends up being represented by an atomic type, so that's a nice improvement.

tbh I haven't checked but I reckon it should. Rustanalyzer says this enum is 1 byte big. And the docs says that whenever it is mappable to an atomic type it will do it.
Otherwise it takes a global lock - but I skimmed the impl: it's more like a pool of fixed number of global locks depending on the address of the thing, so not even that bad (and tbf we don't care here)